kikuchipy/indexing/_refinement.py

```diff
@@ -0,0 +1,1686 @@
+        For the list of possible keyword arguments, see
+        the scipy documentation.
+    trust_region
+        List of how wide the bounds, centered on r and pc,
+        should be for (phi1, Phi, phi2, PCx, PCy, PCz).
+        Only used for methods that support bounds (excluding Powell).
+
+    Returns
+    -------
+        score
+            Highest NCC value.
+        phi1, Phi , phi2
+            The euler angles which gave the highest score, in radians.
+        pcx, pxy, pxz
+            The projection center coordinates which gave the highest
+            score, in the Bruker convention.
+    """
+    phi1_0 = r[..., 0]
+    Phi_0 = r[..., 1]
+    phi2_0 = r[..., 2]
+    eu_x0 = np.array((phi1_0, Phi_0, phi2_0))
+
+    args = (exp,) + pre_args
+
+    full_x0 = np.concatenate((eu_x0, pc), axis=None)
+
+    if method.__name__ == "minimize":
+        soln = method(
+            _full_objective_function_euler,
+            x0=full_x0,
+            args=args,
+            **method_kwargs,
+        )
+    elif method.__name__ == "differential_evolution":
+        soln = method(
+            _full_objective_function_euler,
+            bounds=[
+                (full_x0[0] - trust_region[0], full_x0[0] + trust_region[0]),
+                (full_x0[1] - trust_region[1], full_x0[1] + trust_region[1]),
+                (full_x0[2] - trust_region[2], full_x0[2] + trust_region[2]),
+                (full_x0[3] - trust_region[3], full_x0[3] + trust_region[3]),
+                (full_x0[4] - trust_region[4], full_x0[4] + trust_region[4]),
+                (full_x0[5] - trust_region[5], full_x0[5] + trust_region[5]),
+            ],
+            args=args,
+            **method_kwargs,
+        )
+    elif method.__name__ == "dual_annealing":
+        soln = method(
+            _full_objective_function_euler,
+            bounds=[
+                (full_x0[0] - trust_region[0], full_x0[0] + trust_region[0]),
+                (full_x0[1] - trust_region[1], full_x0[1] + trust_region[1]),
+                (full_x0[2] - trust_region[2], full_x0[2] + trust_region[2]),
+                (full_x0[3] - trust_region[3], full_x0[3] + trust_region[3]),
+                (full_x0[4] - trust_region[4], full_x0[4] + trust_region[4]),
+                (full_x0[5] - trust_region[5], full_x0[5] + trust_region[5]),
+            ],
+            args=args,
+            **method_kwargs,
+        )
+    elif method.__name__ == "basinhopping":
+        method_kwargs["minimizer_kwargs"]["args"] = args
+        soln = method(
+            _full_objective_function_euler,
+            x0=full_x0,
+            **method_kwargs,
+        )
+
+    score = 1 - soln.fun
+    phi1 = soln.x[0]
+    Phi = soln.x[1]
+    phi2 = soln.x[2]
+    pcx = soln.x[3]
+    pxy = soln.x[4]
+    pxz = soln.x[5]
+
+    return (score, phi1, Phi, phi2, pcx, pxy, pxz)
+
+
+def _refine_pc_solver(
+    exp: np.ndarray,
+    r: np.ndarray,
+    pc: np.ndarray,
+    method: type(lambda x: None),
```

Comment:
What does this do? Is "callable" an alternative? If not, perhaps just leave "method" without a type hint...

Reply:
```python
>>> type(lambda x: None)
function
>>> type(getattr(scipy.optimize, "minimize"))
function
```
Callable is an alternative, but perhaps this is more precise?

Reply:
I'm just thinking about how it looks in the Sphinx generated documentation. It might be more precise, but I would argue it is less readable.

Reply:
That is true!